cudamapper/src/overlapper.cpp

```diff
@@ -203,5 +204,39 @@ void Overlapper::post_process_overlaps(std::vector<Overlap>& overlaps)
         overlaps.push_back(fused_overlap);
     }
 }
+
+void Overlapper::rescue_overlap_ends(std::vector<Overlap>& overlaps,
+                                     const io::FastaParser& query_parser,
+                                     const io::FastaParser& target_parser,
+                                     int max_extension,
+                                     float required_similarity)
+{
+
+    // Loop over all overlaps
+    // For each overlap, retrieve the read sequence and
+    // check the similarity of the overlapping head and tail sections (matched for length)
+    // If they are more than or equal to <required_similarity> similar, extend the overlap start/end fields by <max_extension> basepairs.
+
+    for (auto& overlap : overlaps)
+    {
+        std::string query_sequence  = query_parser.get_sequence_by_id(overlap.query_read_id_).seq;
+        std::string target_sequence = target_parser.get_sequence_by_id(overlap.target_read_id_).seq;
+        // TODO: reverse complement target if match is on '-' relative strand.
+
+        // overlap rescue at "head" (i.e., "left-side") of overlap
+        std::size_t query_rescue_head_start  = std::min(static_cast<int32_t>(0), static_cast<int32_t>(overlap.query_start_position_in_read_) - max_extension);
```

Comment:
@ahehn-nv Here's another example where safe `cast_to_signed()` could be useful.

It could be implemented similarly to https://github.com/clara-parabricks/ClaraGenomicsAnalysis/blob/dba8fe0cedc30d181d8057cf2d7940830cae1f89/common/base/include/claragenomics/utils/signed_integer_utils.hpp#L22

Reply:
First, concerning this function:
I think here is a bug. I haven't looked at the PR in detail, but I guess the idea was to make sure that `query_rescure_head_start` is greater-or-equal than zero. It actually makes sure that the value is either zero or negative (which will then become a huge unsigned value). So it should be `std::max`, right?

Concerning the `cast_to_signed`, I can see how this would help here, but I actually think that the problem is further down...
The example shows well, why I would like to avoid unsigned variables basically everywhere and use signed integers for everything that represents a number (including indices and sizes).
Fortunately, Eric was aware that `overlap.query_start_position_in_read_` was an unsigned variable and that he had to cast it to signed first in order for
```
max(0, position - max_extension);
```
to work. If he had been a little less careful, this would have been a pretty bad bug.

So, @mimaric @vellamike: Could we make `position_in_read_t` = `std::int32_t;` (i.e. `overlap.query_start_position_in_read_`) instead of `uint32_t`? Or do we expect reads of a length between 2x10^9 and 4x10^9 bps?

Then `query_rescue_head_start` could be of type `position_in_read_t` as well. Correct?

Reply:
> the idea was to make sure that query_rescure_head_start is greater-or-equal than zero

That's correct, and this was indeed a bug (currently fixing).

> Fortunately, @edawson was aware that overlap.query_start_position_in_read_ was an unsigned variable

I have to credit the compiler here for knowing they were incompatible types. I don't think we'd see a position in the 2-4x10^9 (i.e., 2-4 gigabase) range unless we were doing genome-genome overlapping on a very large genome. I think it's fine to have unsigned positions but I'd like to add a note in the contributing docs reflecting that all position values are unsigned if so.

Reply:
I'm ok with turning everything to signed and casting to unsigned if needed (e.g. getting container size). We'll just have to create a PR for that...

Reply:
Addressed in 2b78224. I switched to `position_in_read_t_` which I think is the proper type to use here. I also fixed the bug described by @ahehn-nv .